language/move-binary-format/src/binary_views.rs

```diff
@@ -199,35 +198,158 @@ impl<'a> BinaryIndexedView<'a> {
+    pub fn module_id_for_handle(&self, module_handle: &ModuleHandle) -> ModuleId {
         ModuleId::new(
             *self.address_identifier_at(module_handle.address),
             self.identifier_at(module_handle.name).to_owned(),
         )
     }
 
-    pub(crate) fn self_id(&self) -> Option<ModuleId> {
+    pub fn self_id(&self) -> Option<ModuleId> {
         match self {
             BinaryIndexedView::Module(m) => Some(m.self_id()),
             BinaryIndexedView::Script(_) => None,
         }
     }
 
-    pub(crate) fn version(&self) -> u32 {
+    pub fn version(&self) -> u32 {
         match self {
             BinaryIndexedView::Module(module) => module.version(),
             BinaryIndexedView::Script(script) => script.version(),
         }
     }
 }
 
+// A `BinaryIndexedViewMut` provides table indexed access for `CompiledModuleMut`
+// and `CompiledScriptMut` -- it is the "mutable" counterpart to `BinaryIndexedView`,
+// with one distinction: this class provides access to the underlying collections
+// within the script or module for bounds checking. Accessing collections that cannot exist on
+// a script results in an empty slice, not an error.
+pub enum BinaryIndexedViewMut<'a> {
+    Module(&'a CompiledModuleMut),
+    Script(&'a CompiledScriptMut),
+}
+
+impl<'a> BinaryIndexedViewMut<'a> {
+    pub fn as_module(&self) -> PartialVMResult<&'a CompiledModuleMut> {
+        match self {
+            BinaryIndexedViewMut::Module(module) => Ok(&module),
+            BinaryIndexedViewMut::Script(_) => {
+                Err(PartialVMError::new(StatusCode::INVALID_OPERATION_IN_SCRIPT))
+            }
+        }
+    }
+
+    pub fn module_handles(&self) -> &[ModuleHandle] {
+        match self {
+            BinaryIndexedViewMut::Module(module) => &module.module_handles,
+            BinaryIndexedViewMut::Script(script) => &script.module_handles,
+        }
+    }
+
+    pub fn signatures(&self) -> &[Signature] {
+        match self {
+            BinaryIndexedViewMut::Module(module) => &module.signatures,
+            BinaryIndexedViewMut::Script(script) => &script.signatures,
+        }
+    }
+
+    pub fn constant_pool(&self) -> &[Constant] {
+        match self {
+            BinaryIndexedViewMut::Module(module) => &module.constant_pool,
+            BinaryIndexedViewMut::Script(script) => &script.constant_pool,
+        }
+    }
+
+    pub fn struct_handles(&self) -> &[StructHandle] {
+        match self {
+            BinaryIndexedViewMut::Module(module) => &module.struct_handles,
+            BinaryIndexedViewMut::Script(script) => &script.struct_handles,
+        }
+    }
+
+    pub fn function_handles(&self) -> &[FunctionHandle] {
+        match self {
+            BinaryIndexedViewMut::Module(module) => &module.function_handles,
+            BinaryIndexedViewMut::Script(script) => &script.function_handles,
+        }
+    }
+
+    pub fn field_handles(&self) -> &[FieldHandle] {
```

Comment:
I would make this (and several below) an option, but the point is kind of moot...